src/acquisition/covid_hosp/common/database.py

```diff
@@ -172,21 +172,31 @@ def insert_dataset(self, publication_date, dataframe):
                        tuple(values) +
                        tuple(i[0] for i in self.additional_fields))
 
+    # deal with non/seldomly updated columns used like a fk table (if this database needs it)
     if hasattr(self, 'AGGREGATE_KEY_COLS'):
       ak_cols = self.AGGREGATE_KEY_COLS
 
       # restrict data to just the key columns and remove duplicate rows
       ak_data = dataframe[ak_cols].drop_duplicates()
-      # cast types (bools and NULLables)
+      # cast types
+      dataframe_typemap = {
+        name: dtype
+        for name, _, dtype in dataframe_columns_and_types
+      }
       for col in ak_cols:
-        if col.startswith('is_'): # TODO: this is hacky af
-          ak_data[col] = (ak_data[col] == 'true')
+          def cast_but_sidestep_nans(i):
+            # not the prettiest, but it works to avoid the NaN values that show up in many columns
+            if isinstance(i, float) and math.isnan(i):
+              return None
+            return dataframe_typemap[col](i)
```

Comment:
We also do that here:
https://github.com/cmu-delphi/delphi-epidata/blob/50e40ba964d19f50563f58ddc54e758af768a068/src/acquisition/covid_hosp/common/database.py#L166-L169

should this code get pulled into a function that can be used in both places?

Reply:
also/alternatively: this is a touch on the YAGNI side, since I can't imagine a scenario where a float would become part of an aggregate key. maybe the check is unnecessary?

Reply:
I considered treading down the slippery slope of refactoring/repairing this (and other) code in this method and module, and ultimately decided it was better to keep my changes separate and distinct.  I can do something to remove the duplication you pointed out, but it felt like touching any of that code would be like opening a can of worms...  My previous "hacky af" handling that only really called out the boolean column(s) was an attempt to produce a smaller and easily understandable diff.

Unfortunately, if we are going to apply the types as defined [here](https://github.com/cmu-delphi/delphi-epidata/blob/ccad94a1ba1b16fb246c4a16e03e5c8c8dac3d35/src/acquisition/covid_hosp/facility/database.py#L18), this check is needed in some form or another.  The dataframe passed as an argument to this method has a NaN wherever the input CSV has an empty field, no matter which column that field is from.  If we apply `str()` (for instance) from `dataframe_typemap` to such a NaN, we get the string `'nan'` when we should really have `None` or even just an empty string.  I almost did something like:
```
for col in ak_cols:
    ak_data[col] = list(map(dataframe_typemap[col], ak_data[col].to_numpy(na_value=None)))
```
but that still needs a guard to prevent (for instance) the string `'None'` from appearing in the database where we now have a `NULL`.  Instead of inventing a new and somewhat overly complicated process, i cribbed from what already existed.

I dont love where this ended up, but there were downsides to each of the options: like the "hacky af" thing that i had before, what we have here in the most recent commit, the incomplete thing i just described in the snippet above, or (as you suggested) altering the existing code to factor out the duplicated functionality.  Maybe i should have just done the latter, as it wouldve taken me a lot less time than typing all this up (lolz).

Reply:
so a string column with one blank row in the csv gets converted into a mixed-type column in the pandas df where some rows are strings and some rows are floats?

Reply:
sorta, except theyre all mixed-type columns (more precisely, `dtype: object`) between when the csv is read and here 😬